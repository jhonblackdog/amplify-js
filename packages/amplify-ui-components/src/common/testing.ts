export const pixelThreshold = 10;
