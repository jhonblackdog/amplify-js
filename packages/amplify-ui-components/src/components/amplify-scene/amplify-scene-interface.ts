export interface AmplifySceneError {
  displayText: string;
  error: object;
}
