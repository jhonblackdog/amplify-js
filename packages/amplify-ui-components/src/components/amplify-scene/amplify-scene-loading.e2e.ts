// TODO: Add e2e testing when aws-sdk-v3 branch is stable
describe('amplify-scene e2e:', () => {
  it(`renders`, async () => {
    expect(true).toBe(true);
  });
});
