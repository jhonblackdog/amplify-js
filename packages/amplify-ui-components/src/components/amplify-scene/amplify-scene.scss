.scene-container {
  width: 100%;
  height: 600px;
  position: relative;
}

.scene {
  width: 100%;
  height: 100%;
}

.scene-nar {
  display: flex;
  align-items: center;
  font-size: 100%;
  position: absolute;
  bottom: 0;
  width: 100%;
}

.scene-actions {
  margin: 1em 1em 1em auto;
}
