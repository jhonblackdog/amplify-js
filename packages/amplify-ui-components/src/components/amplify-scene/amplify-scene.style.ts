import { css } from 'emotion';

export const sceneContainer = css`
  width: 100%;
  height: 600px;
  position: relative;
`;

export const scene = css`
  width: 100%;
  height: 100%;
`;

export const sceneBar = css`
  display: flex;
  align-items: center;
  font-size: 100%;
  position: absolute;
  bottom: 0;
  width: 100%;
`;

export const sceneActions = css`
  margin: 1em 1em 1em auto;
`;
