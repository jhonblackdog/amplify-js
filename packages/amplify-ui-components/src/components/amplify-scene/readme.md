# amplify-scene

<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type     | Default     |
| ----------- | ------------ | ----------- | -------- | ----------- |
| `sceneName` | `scene-name` |             | `string` | `undefined` |


## Dependencies

### Used by

 - [amplify-examples](../amplify-examples)

### Depends on

- [amplify-icon-button](../amplify-icon-button)
- [amplify-scene-loading](../amplify-scene-loading)

### Graph
```mermaid
graph TD;
  amplify-scene --> amplify-icon-button
  amplify-scene --> amplify-scene-loading
  amplify-icon-button --> amplify-tooltip
  amplify-icon-button --> amplify-icon
  amplify-examples --> amplify-scene
  style amplify-scene fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
