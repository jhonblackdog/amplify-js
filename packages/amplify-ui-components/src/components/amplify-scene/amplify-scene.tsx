import { Component, Prop, State, h } from '@stencil/core';
import { sceneContainer, scene, sceneBar, sceneActions } from './amplify-scene.style';
import { AmplifySceneError } from './amplify-scene-interface';
// import { ConsoleLogger as Logger } from '@aws-amplify/core';
import XR from '@aws-amplify/xr';

const SCENE_CONTAINER_DOM_ID = 'scene-container-dom-id';
const SCENE_DOM_ID = 'scene-dom-id';

// TODO: Add logger after Stencil PR is merged: https://github.com/ionic-team/stencil/pull/1773
// const logger = new Logger('amplify-scene');

@Component({
  tag: 'amplify-scene',
  shadow: false,
})
export class AmplifyScene {
  /* The name of the scene being downloaded and rendered */
  @Prop() sceneName: string;
  /* Whether or not the scene is loading */
  @State() loading: boolean = false;
  /* Value between 0-100 that represents the scene's percentage loaded */
  @State() loadPercentage: number = 0;
  /* Whether or not to show a tooltip to enable audio due to browser security requiring it to be disabled on load*/
  @State() showEnableAudio: boolean = false;
  /* Whether or not the scene volume is muted or not */
  @State() muted: boolean = false;
  /* Whether or not the scene's VR presentation is active */
  @State() isVRPresentationActive: boolean = false;
  /* Scene error object */
  @State() sceneError: AmplifySceneError | null = null;
  /* Whether or not the scene is fullscreen */
  @State() isFullscreen: boolean = false;

  async componentDidLoad() {
    this.listenForFullscreenChange();
    await this.loadAndSetupScene(this.sceneName, SCENE_DOM_ID);
  }

  listenForFullscreenChange() {
    document.onfullscreenchange = () => this.onFullscreenChange();
  }

  async loadAndSetupScene(sceneName, sceneDomId) {
    this.loading = true;
    const sceneOptions = {
      progressCallback: progress => {
        this.loadPercentage = progress * 100;
      },
    };
    try {
      await XR.loadScene(sceneName, sceneDomId, sceneOptions);
    } catch (e) {
      this.sceneError = {
        displayText: 'Failed to load scene',
        error: e,
      };
      return;
    }

    XR.start(sceneName);

    this.muted = XR.isMuted(sceneName);
    this.isVRPresentationActive = XR.isVRPresentationActive(sceneName);
    this.loading = false;

    XR.onSceneEvent(sceneName, 'AudioEnabled', () => (this.showEnableAudio = false));
    XR.onSceneEvent(sceneName, 'AudioDisabled', () => (this.showEnableAudio = true));
  }

  setMuted(muted) {
    if (this.showEnableAudio) {
      XR.enableAudio(this.sceneName);
      this.showEnableAudio = false;
    }

    XR.setMuted(this.sceneName, muted);
    this.muted = muted;
  }

  onFullscreenChange() {
    this.isFullscreen = document.fullscreenElement !== null;
  }

  async minimize() {
    if (document['exitFullscreen']) {
      document['exitFullscreen']();
    } else if (document['mozCancelFullScreen']) {
      document['mozCancelFullScreen']();
    } else if (document['webkitExitFullscreen']) {
      document['webkitExitFullscreen']();
    }
  }

  async maximize() {
    const sceneDomElement = document.getElementById(SCENE_CONTAINER_DOM_ID);
    await sceneDomElement.requestFullscreen();
  }

  toggleVRPresentation() {
    try {
      if (this.isVRPresentationActive) {
        XR.exitVR(this.sceneName);
      } else {
        XR.enterVR(this.sceneName);
      }
    } catch (e) {
      // logger.error('Unable to start/stop WebVR System: ' + e.message);
      return;
    }
    this.isVRPresentationActive = !this.isVRPresentationActive;
  }

  render() {
    let muteButton;
    let enterOrExitVRButton;
    let screenSizeButton;

    if (XR.isSceneLoaded(this.sceneName)) {
      if (this.showEnableAudio) {
        muteButton = (
          <amplify-icon-button
            name="sound-mute"
            tooltip="The scene is muted. Click to unmute."
            onClick={() => this.setMuted(false)}
            autoShowTooltip
          />
        );
      } else if (XR.isMuted(this.sceneName)) {
        muteButton = <amplify-icon-button name="sound-mute" tooltip="Unmute" onClick={() => this.setMuted(false)} />;
      } else {
        muteButton = <amplify-icon-button name="sound" tooltip="Mute" onClick={() => this.setMuted(true)} />;
      }

      if (XR.isVRCapable(this.sceneName)) {
        if (this.isVRPresentationActive) {
          // logger.info('VR Presentation Active');
          enterOrExitVRButton = (
            <amplify-icon-button name="exit-vr" tooltip="Exit VR" onClick={() => this.toggleVRPresentation()} />
          );
        } else {
          // logger.info('VR Presentation Inactive');
          enterOrExitVRButton = (
            <amplify-icon-button name="enter-vr" tooltip="Enter VR" onClick={() => this.toggleVRPresentation()} />
          );
        }
      }

      if (this.isFullscreen) {
        screenSizeButton = (
          <amplify-icon-button name="minimize" tooltip="Exit Fullscreen" onClick={() => this.minimize()} />
        );
      } else {
        screenSizeButton = <amplify-icon-button name="maximize" tooltip="Fullscreen" onClick={() => this.maximize()} />;
      }
    }

    return (
      <div id={SCENE_CONTAINER_DOM_ID} class={sceneContainer}>
        <div id={SCENE_DOM_ID} class={scene}>
          {this.loading ? (
            <amplify-scene-loading
              scene-name={this.sceneName}
              load-percentage={this.loadPercentage}
              scene-error={this.sceneError}
            />
          ) : null}
        </div>

        <div class={sceneBar}>
          <span class={sceneActions}>
            {muteButton}
            {enterOrExitVRButton}
            {screenSizeButton}
          </span>
        </div>
      </div>
    );
  }
}
