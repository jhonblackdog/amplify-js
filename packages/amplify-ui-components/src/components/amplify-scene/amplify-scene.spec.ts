import { newSpecPage } from '@stencil/core/testing';
import { AmplifyScene } from './amplify-scene';
import { XR } from 'aws-amplify';

describe('amplify-scene spec:', () => {
  describe('Component logic ->', () => {
    let amplifyScene;

    beforeEach(() => {
      amplifyScene = new AmplifyScene();
      amplifyScene.loadAndSetupScene = jest.fn();
    });

    it('`sceneName` should be undefined by default', () => {
      expect(amplifyScene.sceneName).toBeUndefined();
    });

    it('`loading` should be false by default', () => {
      expect(amplifyScene.loading).toBeFalsy();
    });
  });
  describe('Render logic ->', () => {
    beforeEach(() => {
      XR.loadScene = jest.fn();
      XR.isSceneLoaded = jest.fn(() => false);
      XR.isMuted = jest.fn(() => false);
      XR.isVRPresentationActive = jest.fn();
      XR.isVRCapable = jest.fn();
      XR.getSceneController = jest.fn();
    });
    it(`should render loading overlay by default`, async () => {
      const page = await newSpecPage({
        components: [AmplifyScene],
        html: `<amplify-scene sceneName={'scene1'} />`,
      });
      expect(page.root).toMatchSnapshot();
    });
    it(`should render scene action bar when 'isSceneLoaded' is true`, async () => {
      XR.isSceneLoaded = jest.fn(() => true);
      const page = await newSpecPage({
        components: [AmplifyScene],
        html: `<amplify-scene sceneName={'scene1'} />`,
      });
      expect(page.root).toMatchSnapshot();
    });
  });
});
