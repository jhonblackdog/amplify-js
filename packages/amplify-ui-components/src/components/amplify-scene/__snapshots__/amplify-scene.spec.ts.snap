// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-scene spec: Render logic -> should render loading overlay by default 1`] = `
<amplify-scene scenename="{'scene1'}">
  <div class="css-etfphl" id="scene-container-dom-id">
    <div class="css-s08p0c" id="scene-dom-id">
      <amplify-scene-loading load-percentage="0"></amplify-scene-loading>
    </div>
    <div class="css-xdndw7">
      <span class="css-31nfqj"></span>
    </div>
  </div>
</amplify-scene>
`;

exports[`amplify-scene spec: Render logic -> should render scene action bar when 'isSceneLoaded' is true 1`] = `
<amplify-scene scenename="{'scene1'}">
  <div class="css-etfphl" id="scene-container-dom-id">
    <div class="css-s08p0c" id="scene-dom-id">
      <amplify-scene-loading load-percentage="0"></amplify-scene-loading>
    </div>
    <div class="css-xdndw7">
      <span class="css-31nfqj">
        <amplify-icon-button name="sound" tooltip="Mute"></amplify-icon-button>
        <amplify-icon-button name="maximize" tooltip="Fullscreen"></amplify-icon-button>
      </span>
    </div>
  </div>
</amplify-scene>
`;
